### PR TITLE
Fixes for POST policy checks and the x-ignore implementation

### PR DESCRIPTION
**Pull request by marktheunissen**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description

Related PR: https://github.com/minio/minio-go/pull/1996

- `X-AWS-Ignore-...` was incorrectly implemented according to S3 docs.
- `Content-Encoding` should not be an ignored form field.
- The checksum form fields should not be ignored either.
- Simplify checkPostPolicy and add more tests

## Motivation and Context

We're not correctly implementing S3 spec, and there are bugs in the implementation we do have.

Our current implementation allows form values like `X-Amz-Ignore-Foo` to work as directives that tell MinIO to ignore `Foo` form value, so you could provide `Foo` and `X-Amz-Ignore-Foo` together, and then we would not check that `Foo` appears in the policy.  This implementation was probably just a misunderstanding or a misread of what the actual behavior should be, as specified here: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html

`X-Amz-Ignore..` is not mentioned at all in AWS docs that I can find. Only `X-Ignore-...`, and the way it works is that any form key starting with `X-Ignore-...` will be ignored - it does not work as a directive that affects other form keys.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/postpolicyform.go [hunk 1]
```diff
@@ -272,96 +260,92 @@ func checkPolicyCond(op string, input1, input2 string) bool {
...
+// S3 docs: "Each form field that you specify in a form (except x-amz-signature, file, policy, and field names
+// that have an x-ignore- prefix) must appear in the list of conditions."
+// https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html
+// keyInPolicyExceptions - list of keys that, when present in the form, can be missing in the conditions of the policy.
+var keyInPolicyExceptions = map[string]bool{
+	xhttp.AmzSignature: true,
+	"File":             true,
+	"Policy":           true,
+
+	// MinIO specific exceptions to the general S3 rule above.
```
Comment (marktheunissen) [inline cmd/postpolicyform.go:272]: I'm removing the check `strings.HasPrefix(key, encrypt.SseGenericHeader)` below, and instead listing out the SSE related header exceptions here.

Please check my assumption that ignoring these SSE headers is still the correct course of action.

File: cmd/postpolicyform_test.go [hunk 2]
```diff
@@ -78,83 +78,221 @@ func TestParsePostPolicyForm(t *testing.T) {
...
-			t.Fatal(err)
-		}
-
-		postPolicyForm, err := parsePostPolicyForm(bytes.NewReader(policyBytes))
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		err = checkPostPolicy(formValues, postPolicyForm)
-		if err != nil && tt.expectedErr != nil && err.Error() != tt.expectedErr.Error() {
```
Comment (marktheunissen) [inline cmd/postpolicyform_test.go:156]: This was masking the fact that at least one of the existing tests was actually failing.

File: go.mod [hunk 3]
```diff
@@ -265,3 +265,5 @@ require (
 	google.golang.org/protobuf v1.35.1 // indirect
 
 )
+
+replace github.com/minio/minio-go/v7 => github.com/marktheunissen/minio-go/v7 v7.0.0-20241121022745-728929652a43
```
Comment (marktheunissen) [inline go.mod]: Do not merge until this is removed.

**Comment by marktheunissen** [comment]
Tests will fail until the `mint` image is updated with this: https://github.com/minio/minio-go/pull/1996

**Comment by klauspost** [comment]
Failing mint test: https://github.com/minio/minio-go/blob/master/functional_tests.go#L5635

And regular test:

```
2024-11-21T04:57:21.6565040Z --- FAIL: TestPostPolicyForm (0.00s)
....
2024-11-21T04:57:21.6578393Z     --- FAIL: TestPostPolicyForm/unknown_key_Content-Encoding_is_error_as_it_does_not_appear_in_policy (0.00s)
```

**Comment by marktheunissen** [comment]
@klauspost Thanks, I fixed the regular test, and yes, Mint tests will fail until the mint image is updated with this: https://github.com/minio/minio-go/pull/1996

What procedure would you prefer to follow here?

**Comment by klauspost** [comment]
Let's get upstream merged.

File: cmd/postpolicyform.go [hunk 4]
```diff
@@ -272,96 +260,92 @@ func checkPolicyCond(op string, input1, input2 string) bool {
...
-		}
-	}
-	// map to store the metadata
-	metaMap := make(map[string]string)
-	for _, policy := range postPolicyForm.Conditions.Policies {
-		if strings.HasPrefix(policy.Key, "$x-amz-meta-") {
-			formCanonicalName := http.CanonicalHeaderKey(strings.TrimPrefix(policy.Key, "$"))
-			metaMap[formCanonicalName] = policy.Value
 		}
+		mustFindInPolicy[key] = values
```
Comment (klauspost) [inline cmd/postpolicyform.go:297]: Are form keys always canonical? If not, the keys should probably be made canonical?

File: cmd/postpolicyform.go [hunk 5]
```diff
@@ -272,96 +260,92 @@ func checkPolicyCond(op string, input1, input2 string) bool {
...
-	checkHeader := map[string][]string{}
-	ignoreKeys := map[string]bool{}
-	for key, value := range formValues {
-		switch {
-		case ignoreKeys[key], postPolicyIgnoreKeys[key], strings.HasPrefix(key, encrypt.SseGenericHeader):
+
+	// mustFindInPolicy is a map to list all the keys that we must find in the policy as
+	// we process it below. At the end of checkPostPolicy function, if any key is left in
+	// this map, that's an error.
+	mustFindInPolicy := map[string][]string{}
```
Comment (klauspost) [inline cmd/postpolicyform.go]:
```suggestion
	mustFindInPolicy := make(map[string][]string, len(formvalues))
```

File: go.mod (hunk 3 above)
Comment (klauspost) [reply go.mod]: Can be removed now, correct?

**Review by klauspost — COMMENTED** [review]
Just a few overall Qs.

File: cmd/postpolicyform.go (hunk 4 above)
Comment (marktheunissen) [reply cmd/postpolicyform.go:297]: Yes they are, starting with line 975 here: https://github.com/minio/minio/blob/master/cmd/bucket-handlers.go#L974-L1068.

I'll add a note in this section of the code as a comment.

**Review by klauspost — APPROVED** [review]
LGTM

**Comment by marktheunissen** [comment]
Blocked currently by this failing mint test, which is unrelated:
```
{
  "time": "2024-12-04T05:45:17.485590419Z",
  "level": "ERROR",
  "name": "minio-go: testGetObjectAttributes",
  "duration": 474,
  "function": "GetObjectAttributes(ctx, bucketName, objectName, opts)",
  "args": {
    "bucketName": "minio-go-test-versioned-wcioaf",
    "objectName": "file1",
    "opts": "minio.ObjectAttributesOptions{}"
  },
  "status": "FAIL",
  "message": "Validating GetObjectsAttributes response failed, table test: versioned",
  "error": "Full object checksum not found"
}
```

I started discussion on the fix to the above test here: https://github.com/minio/minio/pull/20732

**Review by harshavardhana — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]
PR already merged in aistor

**Pull request merged**